include/alpaka/core/Unreachable.hpp

```diff
@@ -0,0 +1,30 @@
+/* Copyright 2022 Jan Stephan
+ *
+ * This file is part of alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <alpaka/core/BoostPredef.hpp>
+
+//! Before CUDA 11.5 nvcc is unable to correctly identify return statements in 'if constexpr' branches. It will issue
+//! a false warning about a missing return statement unless it is told that the following code section is unreachable.
+//!
+//! \param x A dummy value for the expected return type of the calling function.
+#if(BOOST_COMP_NVCC && BOOST_ARCH_PTX)
+#    if BOOST_LANG_CUDA >= BOOST_VERSION_NUMBER(11, 3, 0)
+#        define ALPAKA_UNREACHABLE(...) __builtin_unreachable()
+#    else
+#        define ALPAKA_UNREACHABLE(...) return __VA_ARGS__
+#    endif
+#elif BOOST_COMP_MSVC
+#    define ALPAKA_UNREACHABLE(...) __assume(false)
+#elif BOOST_COMP_GNUC || BOOST_COMP_CLANG || BOOST_COMP_INTEL || BOOST_COMP_PGI
+#    define ALPAKA_UNREACHABLE(...) __builtin_unreachable()
```

Comment:
This does not exactly seem to work for NVHPC/PGI, [e.g.](https://github.com/alpaka-group/alpaka/runs/4790090659?check_suite_focus=true#step:5:3141).

Reply:
So NVHPC warns about `__builtin_unreachable()` being unreachable... *sigh*. I give up, I guess we have to live with that warning until we can remove CUDA < 11. or NVHPC is fixed.

Reply:
If the current version removes the warning for CUDA < 11, then keep it. With NVHPC we do not have to support old versions, i.e. it just moves up with time dropping the old ones.

Reply:
Sorry, there was a typo in my comment (I edited it to say "< CUDA 11.5"). But I'm still hopeful that NVHPC will be fixed sooner than us dropping everything except the current latest CUDA version.

Reply:
It looks like `__builtin_unreachable()` is just not supported by PGI/NVHPC and may never be as it is a GNU buiilt-in.

So, remove PGI here:
```suggestion
#elif BOOST_COMP_GNUC || BOOST_COMP_CLANG || BOOST_COMP_INTEL
#    define ALPAKA_UNREACHABLE(...) __builtin_unreachable()
```
and instead add `--diag_suppress 111 ` [here](https://github.com/alpaka-group/alpaka/blob/develop/cmake/alpakaCommon.cmake#L146) to disable this warning.